gameoflife/src/test/kotlin/com/anaplan/engineering/kazuki/gameoflife/TestConway.kt

```diff
@@ -0,0 +1,303 @@
+import com.anaplan.engineering.kazuki.gameoflife.Conway_Module.mk_Population
+import kotlin.test.Ignore
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+
+
+class TestConway {
+
+    private val block = mk_Population(mk_Point(0, 0), mk_Point(-1, 0), mk_Point(0, -1), mk_Point(-1, -1))
+
+    private val blinker = mk_Population(mk_Point(-1, 0), mk_Point(0, 0), mk_Point(1, 0))
+
+    private val toad = mk_Population(
+        mk_Point(-1, 0), mk_Point(0, 0), mk_Point(1, 0),
+        mk_Point(0, -1), mk_Point(-1, -1), mk_Point(-2, -1)
+    )
+
+    private val beacon = mk_Population(
+        mk_Point(-2, 0), mk_Point(-2, 1), mk_Point(-1, 1), mk_Point(0, -2),
+        mk_Point(1, -2), mk_Point(1, -1)
+    )
+
+    private val pQuad = mk_Population(
+        mk_Point(2, 1), mk_Point(3, 1), mk_Point(3, 2),
+        mk_Point(1, 2), mk_Point(1, 3), mk_Point(2, 3), mk_Point(5, 2),
+        mk_Point(5, 3), mk_Point(6, 3), mk_Point(7, 3), mk_Point(2, 5),
+        mk_Point(3, 5), mk_Point(3, 6), mk_Point(3, 7)
+    )
+
+    private val pulsar = mk_Population(
+        dunion(
+            set(pQuad) { point -> point },
+            set(pQuad) { point -> mk_Point(-point.x, point.y) },
+            set(pQuad) { point -> mk_Point(point.x, -point.y) },
+            set(pQuad) { point -> mk_Point(-point.x, -point.y) }
+        ))
+
+    private val diehard = mk_Population(
+        mk_Point(0, 1), mk_Point(1, 1), mk_Point(1, 0), mk_Point(0, 5),
+        mk_Point(0, 6), mk_Point(0, 7), mk_Point(2, 6)
+    )
+
+    private val glider = mk_Population(
+        mk_Point(1, 0), mk_Point(2, 0), mk_Point(3, 0),
+        mk_Point(3, 1), mk_Point(2, 2)
+    )
+
+    @Test
+    @Ignore
+    fun preconditionTests() {
+        // They should not allow these 0s to be input, as they should be a nat1 number
+        assertFailsWith<PreconditionFailure> { generations(0, block) }
+        assertFailsWith<PreconditionFailure> { periodN(block, 0) }
+        assertFailsWith<PreconditionFailure> { disappearN(block, 0) }
+        assertFailsWith<PreconditionFailure> { disappearNP(block, 0) }
+        assertFailsWith<PreconditionFailure> { gliderN(block, 0, 1) }
+        assertFailsWith<PreconditionFailure> { gliderNP(block, 0, 1) }
+        assertFailsWith<PreconditionFailure> { gliderN(block, 1, 0) }
+        assertFailsWith<PreconditionFailure> { gliderNP(block, 1, 0) }
+        assertFailsWith<PreconditionFailure> { isOffset(block, block, 0) }
+    }
+
+    @Test
+    fun around() {
+        assertEquals(
+            mk_Population(
+                mk_Point(1, 0), mk_Point(1, 1), mk_Point(0, 1), mk_Point(-1, 1),
+                mk_Point(-1, 0), mk_Point(-1, -1), mk_Point(0, -1), mk_Point(1, -1)
+            ),
+            around(mk_Point(0, 0))
+        )
+        assertEquals(true, mk_Point(0, 0) !in around(mk_Point(0, 0)))
+    }
+
+    @Test
+    fun neighbourCount() {
```

Comment:
⛏️ I would have liked to see a test where the point itself and a couple of its neighbours are in the population, but the other neighbours are not.

Reply:
This is the case for the first example, `assertEquals(3, neighbourCount(block, mk_Point(0, 0)))`, as `block` is made up of the four points: { (0,0), (-1,0), (0,-1), (-1,-1) }. I admit it isn't immediately obvious though, so I can replace it with a more explicit version if you'd prefer?

Reply:
Fair enough. It's fine to leave it like that, then.